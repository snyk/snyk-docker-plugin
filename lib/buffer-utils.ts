import * as crypto from "crypto";
import { HashAlgorithm } from "./types";

const HASH_ENCODING = "hex";

export async function bufferToSha1(buffer: Buffer): Promise<string> {
  const hash = crypto.createHash(HashAlgorithm.Sha1);
  const chunkSize = 100 * 1024 * 1024; // 100 MB

  return new Promise((resolve, reject) => {
    try {
      for (let offset = 0; offset < buffer.length; offset += chunkSize) {
        const end = Math.min(offset + chunkSize, buffer.length);
        const chunk = buffer.slice(offset, end);
        hash.update(chunk);
      }

      const digest = hash.digest(HASH_ENCODING);
      resolve(digest);
    } catch (err) {
      reject(err);
    }
  });
}
