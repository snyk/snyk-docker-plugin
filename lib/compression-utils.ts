import { Decompress as ZstdDecompress } from "fzstd";

/**
 * Decompresses zstd-compressed data from a buffer.
 *
 * This is a synchronous buffer-to-buffer decompression utility.
 * For streaming zstd decompression, use the decompressMaybe transform stream.
 *
 * @param compressed Buffer containing zstd-compressed data
 * @returns Decompressed data as a Buffer
 * @throws Error if decompression fails
 */
export function decompressZstd(compressed: Buffer): Buffer {
  const chunks: Buffer[] = [];

  try {
    const decompressor = new ZstdDecompress((data: Uint8Array) => {
      chunks.push(Buffer.from(data));
    });

    decompressor.push(new Uint8Array(compressed), true);

    return Buffer.concat(chunks);
  } catch (error) {
    throw new Error(
      `Zstd decompression failed: ${
        error instanceof Error ? error.message : String(error)
      }`,
    );
  }
}
