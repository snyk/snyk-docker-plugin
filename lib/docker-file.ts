import * as fs from 'fs';
import * as dockerFileParser from 'docker-file-parser';

export { getBaseImageName };

async function getBaseImageName(targetFile?: string):
  Promise<string|undefined> {

  if (!targetFile) {
    return undefined;
  }

  const contents = await readFile(targetFile);
  const commands = dockerFileParser.parse(contents);
  const fromCommands = commands.filter(command => command.name === 'FROM');
  const finalFrom = fromCommands.pop();

  return finalFrom ? String(finalFrom.args) : undefined;
}

async function readFile(path: string) {
  return new Promise((resolve, reject) => {
    fs.readFile(path, 'utf8', (err, data) => {
      return err ? reject(err) : resolve(data);
    });
  }) as Promise<string>;
}
