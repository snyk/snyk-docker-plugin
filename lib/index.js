var subProcess = require('./sub-process');
var fetchSnykDockerAnalyzer = require('./fetch-snyk-docker-analyzer');
var debug = require('debug')('snyk');

module.exports = {
  inspect: inspect,
};

function inspect(root, targetFile, options) {
  const targetImage = root;
  return fetchSnykDockerAnalyzer()
  .then(function (analyzerBinaryPath) {
    return Promise.all([
      getRuntime(),
      getDependencies(analyzerBinaryPath, targetImage),
    ])
    .then(function (result) {
      var metadata = {
        name: 'snyk-docker-plugin',
        runtime: result[0],
        packageManager: result[1].packageManager,
      };
      var package = result[1].package;
      return {
        plugin: metadata,
        package: package,
      };
    });
  })
}

function getRuntime() {
  return subProcess.execute('docker', ['version'])
  .then(function (output) {
    versionMatch = /Version:\s+(.*)\n/.exec(output);
    if (versionMatch) {
      return 'docker ' + versionMatch[1]
    }
    return undefined;
  });
}

function handleCommonErrors(error, targetImage) {
  if (error.indexOf('command not found') !== -1) {
    throw new Error('Snyk docker CLI was not found')
  }
  if (error.indexOf('Cannot connect to the Docker daemon') !== -1) {
    throw new Error('Cannot connect to the Docker daemon. Is the docker'
    + ' daemon running?')
  }
  if ((error.indexOf('Error loading image from docker engine') !== -1) ||
      (error.indexOf('Error performing image analysis') !== -1)) {
    throw new Error('Docker image was not found: ' + targetImage)
  }
  if (error.indexOf('Error getting docker client:') !== -1) {
    throw new Error('Failed getting docker client')
  }
  if (error.indexOf('Error processing image:') !== -1) {
    throw new Error('Failed processing image:' + targetImage)
  }
}

function getDependencies(analyzerBinaryPath, targetImage) {
  var result;
  return subProcess.execute(
    analyzerBinaryPath,
    buildArgs(targetImage)
  )
  .then(function (output) {
    result = parseAnalysisResults(output);
    return buildTree(
      targetImage, result.type, result.depInfosList, result.osRelease);
  })
  .then(function (package) {
    return {
      package: package,
      packageManager: result.type,
    };
  })
  .catch(function (error) {
    if (typeof error === 'string') {
      debug(`Error while running analyser: '${error}'`);
      handleCommonErrors(error, targetImage);
      errorMsg = error;
      errorMatch = /msg="(.*)"/g.exec(errorMsg)
      if (errorMatch) {
        errorMsg = errorMatch[1]
      }
      throw new Error(errorMsg);
    }

    throw error;
  })
}

function parseAnalysisResults(analysisOut) {
  var analysisJson = JSON.parse(analysisOut);

  var analysisResult = analysisJson.results.filter(function (res) {
    return res.Analysis && res.Analysis.length > 0;
  })[0];

  // TODO: if analysisResult is undefined -
  //  throw a nice error, or report 0 deps

  var depType;
  switch (analysisResult.AnalyzeType) {
    case 'Apt': {
      depType = 'deb';
      break;
    }
    default: {
      depType = analysisResult.AnalyzeType.toLowerCase()
    }
  }

  return {
    type: depType,
    depInfosList: analysisResult.Analysis,
    osRelease: analysisJson.osRelease,
  }
}

function buildTree(targetImage, depType, depInfosList, osRelease) {
  var targetSplit = targetImage.split(':');
  var imageName = targetSplit[0];
  var imageVersion = targetSplit[1] ? targetSplit[1] : 'latest';

  var root = {
    name: imageName,
    version: imageVersion,
    dockerOSRelease: osRelease,
    packageFormatVersion: depType + ':0.0.1',
    dependencies: {},
  };

  var depsMap = depInfosList.reduce(function (acc, depInfo) {
    var name = depInfo.Name;
    acc[name] = depInfo;
    return acc;
  }, {});

  var virtualDepsMap = depInfosList.reduce(function (acc, depInfo) {
    var providesNames = depInfo.Provides || [];
    providesNames.forEach(function (name) {
      acc[name] = depInfo;
    });
    return acc;
  }, {});

  var attachDeps = function (depInfos) {
    depInfos.forEach(function (depInfo) {
      var subtree = buildTreeRecurisve(
        depInfo.Name, new Set(), depsMap, virtualDepsMap);
      if (subtree) {
        root.dependencies[subtree.name] = subtree;
      }
    })
  };

  // attach (as direct deps) pkgs not marked auto-installed:
  var manuallyInstalledDeps = depInfosList.filter(function (depInfo) {
    return !depInfo.AutoInstalled;
  })
  attachDeps(manuallyInstalledDeps);

  // attach (as direct deps) pkgs marked as auto-insatalled,
  //  but not dependant upon:
  var notVisitedDeps = depInfosList.filter(function (depInfo) {
    var depName = depInfo.Name;
    return !(depsMap[depName]._visited);
  })
  attachDeps(notVisitedDeps);

  return root;
}

function buildTreeRecurisve(
    depName, ancestors, depsMap, virtualDepsMap) {
  var depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return null;
  }

  var fullName = depFullName(depInfo);
  if (ancestors.has(fullName)) {
    return null;
  }

  const tree = {
    name: fullName,
    version: depInfo.Version,
  };

  //TODO: if we do this we can avoid tracking ancestors
  if (depInfo._visited) {
    return tree;
  }
  depInfo._visited = true;

  var newAncestors = (new Set(ancestors)).add(fullName);

  var deps = depInfo.Deps || {};
  Object.keys(deps).forEach(function (depName) {
    var subTree = buildTreeRecurisve(
      depName, newAncestors, depsMap, virtualDepsMap);
    if (subTree) {
      if (!tree.dependencies) {
        tree.dependencies = {};
      }
      tree.dependencies[subTree.name] = subTree;
    }
  });

  return tree;
}

function depFullName(depInfo) {
  var fullName = depInfo.Name;
  if (depInfo.Source) {
    fullName = depInfo.Source + '/' + fullName;
  }
  return fullName;
}

function buildArgs(targetImage) {
  var args = ['analyze', targetImage];
  return args;
}
