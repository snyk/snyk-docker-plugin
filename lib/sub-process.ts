import * as childProcess from "child_process";
import { quoteAll } from "shescape/stateless";

export { execute, CmdOutput };
interface CmdOutput {
  stdout: string;
  stderr: string;
}

function execute(
  command: string,
  args: string[],
  options?,
): Promise<CmdOutput> {
  const spawnOptions: any = { shell: true, env: process.env };
  if (options && options.cwd) {
    spawnOptions.cwd = options.cwd;
  }
  args = quoteAll(args, spawnOptions);

  // Before spawning an external process, we look if we need to restore the system proxy configuration,
  // which overides the cli internal proxy configuration.
  if (process.env.SNYK_SYSTEM_HTTP_PROXY !== undefined) {
    spawnOptions.env.HTTP_PROXY = process.env.SNYK_SYSTEM_HTTP_PROXY;
  }
  if (process.env.SNYK_SYSTEM_HTTPS_PROXY !== undefined) {
    spawnOptions.env.HTTPS_PROXY = process.env.SNYK_SYSTEM_HTTPS_PROXY;
  }
  if (process.env.SNYK_SYSTEM_NO_PROXY !== undefined) {
    spawnOptions.env.NO_PROXY = process.env.SNYK_SYSTEM_NO_PROXY;
  }

  return new Promise((resolve, reject) => {
    let stdout = "";
    let stderr = "";

    const proc = childProcess.spawn(command, args, spawnOptions);
    proc.stdout.on("data", (data) => {
      stdout = stdout + data;
    });
    proc.stderr.on("data", (data) => {
      stderr = stderr + data;
    });

    proc.on("close", (code) => {
      const output = { stdout, stderr };
      if (code !== 0) {
        return reject(output);
      }
      resolve(output);
    });
  }) as Promise<CmdOutput>;
}
