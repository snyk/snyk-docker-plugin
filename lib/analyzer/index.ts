import * as Debug from "debug";
import { Docker, DockerOptions } from "../docker";
import * as dockerFile from "../docker-file";
import * as apkAnalyzer from "./apk-analyzer";
import * as aptAnalyzer from "./apt-analyzer";
import * as binariesAnalyzer from "./binaries-analyzer";
import * as imageInspector from "./image-inspector";
import * as osReleaseDetector from "./os-release-detector";
import * as rpmAnalyzer from "./rpm-analyzer";

export { analyze };

const debug = Debug("snyk");

const extractActions = [
  ...aptAnalyzer.APT_PKGPATHS,
  ...apkAnalyzer.APK_PKGPATHS,
  ...osReleaseDetector.OS_VERPATHS,
].map((p) => {
  return {
    name: "txt",
    pattern: p,
  };
});

async function analyze(
  targetImage: string,
  dockerfileAnalysis?: dockerFile.DockerFileAnalysis,
  options?: DockerOptions,
) {
  const docker = new Docker(targetImage, options);

  await docker.scanStaticalyIfNeeded(extractActions);

  const [imageInspection, osRelease] = await Promise.all([
    imageInspector.detect(docker),
    osReleaseDetector.detect(docker, dockerfileAnalysis),
  ]);

  const results = await Promise.all([
    apkAnalyzer.analyze(docker),
    aptAnalyzer.analyze(docker),
    rpmAnalyzer.analyze(docker),
  ]).catch((err) => {
    debug(`Error while running analyzer: '${err.stderr}'`);
    throw new Error("Failed to detect installed OS packages");
  });

  const { installedPackages, pkgManager } = getInstalledPackages(
    results as any[],
  );
  let binaries;
  try {
    binaries = await binariesAnalyzer.analyze(
      targetImage,
      installedPackages,
      pkgManager,
      options,
    );
  } catch (err) {
    debug(`Error while running binaries analyzer: '${err}'`);
    throw new Error("Failed to detect binaries versions");
  }

  return {
    imageId: imageInspection.Id,
    osRelease,
    results,
    binaries,
    imageLayers: imageInspection.RootFS && imageInspection.RootFS.Layers,
  };
}

function getInstalledPackages(
  results: any[],
): { installedPackages: string[]; pkgManager?: string } {
  const dockerAnalysis = results.find((res) => {
    return res.Analysis && res.Analysis.length > 0;
  });

  if (!dockerAnalysis) {
    return { installedPackages: [] };
  }
  const installedPackages = dockerAnalysis.Analysis.map((pkg) => pkg.Name);
  let pkgManager = dockerAnalysis.AnalyzeType;
  if (pkgManager) {
    pkgManager = pkgManager.toLowerCase();
  }
  return { installedPackages, pkgManager };
}
