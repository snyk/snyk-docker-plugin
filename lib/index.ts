const debug = require('debug')('snyk');

import * as analyzer from './analyzer';
import * as subProcess from './sub-process';
import * as dockerFile from './docker-file';

export {
  inspect,
};

function inspect(root: string, targetFile?: string, options?: any) {
  const targetImage = root;
  return Promise.all([
    getRuntime(),
    getDependencies(targetImage),
    dockerFile.getBaseImageName(targetFile),
  ])
    .then((result) => {
      const metadata = {
        name: 'snyk-docker-plugin',
        runtime: result[0],
        packageManager: result[1].packageManager,
        dockerImageId: result[1].imageId,
      };
      const pkg: any = result[1].package;
      pkg.docker = pkg.docker || {};
      pkg.docker.baseImage = result[2];
      return {
        plugin: metadata,
        package: pkg,
      };
    });
}

function getRuntime() {
  return subProcess.execute('docker', ['version'])
    .then((output) => {
      const versionMatch = /Version:\s+(.*)\n/.exec(output);
      if (versionMatch) {
        return 'docker ' + versionMatch[1];
      }
      return undefined;
    });
}

function handleCommonErrors(error, targetImage: string) {
  if (error.indexOf('command not found') !== -1) {
    throw new Error('Snyk docker CLI was not found');
  }
  if (error.indexOf('Cannot connect to the Docker daemon') !== -1) {
    throw new Error('Cannot connect to the Docker daemon. Is the docker'
    + ' daemon running?');
  }
  const ERROR_LOADING_IMAGE_STR = 'Error loading image from docker engine:';
  if (error.indexOf(ERROR_LOADING_IMAGE_STR) !== -1) {
    if (error.indexOf('reference does not exist') !== -1) {
      throw new Error(
        `Docker image was not found locally: ${targetImage}`);
    }
    if (error.indexOf('permission denied while trying to connect') !== -1) {
      let errString = error.split(ERROR_LOADING_IMAGE_STR)[1];
      errString = (errString || '').slice(0, -2); // remove trailing \"
      throw new Error(
        'Permission denied connecting to docker daemon. ' +
        'Please make sure user has the required permissions. ' +
        'Error string: ' + errString);
    }
  }
  if (error.indexOf('Error getting docker client:') !== -1) {
    throw new Error('Failed getting docker client');
  }
  if (error.indexOf('Error processing image:') !== -1) {
    throw new Error('Failed processing image:' + targetImage);
  }
}

function getDependencies(targetImage: string) {
  let result;
  return analyzer.analyze(targetImage)
    .then((output) => {
      result = parseAnalysisResults(output);
      return buildTree(
        targetImage, result.type, result.depInfosList, result.targetOS);
    })
    .then((pkg) => {
      return {
        package: pkg,
        packageManager: result.type,
        imageId: result.imageId,
      };
    })
    .catch((error) => {
      if (typeof error === 'string') {
        debug(`Error while running analyzer: '${error}'`);
        handleCommonErrors(error, targetImage);
        let errorMsg = error;
        const errorMatch = /msg="(.*)"/g.exec(errorMsg);
        if (errorMatch) {
          errorMsg = errorMatch[1];
        }
        throw new Error(errorMsg);
      }
      throw error;
    });
}

function parseAnalysisResults(analysisJson) {
  const analysisResult = analysisJson.results.filter((res) => {
    return res.Analysis && res.Analysis.length > 0;
  })[0];

  if (!analysisResult) {
    throw new Error(
      'Failed to detect a supported Linux package manager (deb/rpm/apk)');
  }

  let depType;
  switch (analysisResult.AnalyzeType) {
    case 'Apt': {
      depType = 'deb';
      break;
    }
    default: {
      depType = analysisResult.AnalyzeType.toLowerCase();
    }
  }

  return {
    imageId: analysisJson.imageId,
    targetOS: analysisJson.osRelease,
    type: depType,
    depInfosList: analysisResult.Analysis,
  };
}

function buildTree(targetImage, depType, depInfosList, targetOS) {
  const targetSplit = targetImage.split(':');
  const imageName = targetSplit[0];
  const imageVersion = targetSplit[1] ? targetSplit[1] : 'latest';

  const root = {
    name: imageName,
    version: imageVersion,
    targetOS,
    packageFormatVersion: depType + ':0.0.1',
    dependencies: {},
  };

  const depsMap = depInfosList.reduce((acc, depInfo) => {
    const name = depInfo.Name;
    acc[name] = depInfo;
    return acc;
  }, {});

  const virtualDepsMap = depInfosList.reduce((acc, depInfo) => {
    const providesNames = depInfo.Provides || [];
    providesNames.forEach((name) => {
      acc[name] = depInfo;
    });
    return acc;
  }, {});

  const depsCounts = {};
  depInfosList.forEach((depInfo) => {
    countDepsRecursive(
      depInfo.Name, new Set(), depsMap, virtualDepsMap, depsCounts);
  });
  const DEP_FREQ_THRESHOLD = 100;
  const tooFrequentDepNames = Object.keys(depsCounts)
    .filter((depName) => {
      return depsCounts[depName] > DEP_FREQ_THRESHOLD;
    });

  const attachDeps = (depInfos) => {
    const depNamesToSkip = new Set(tooFrequentDepNames);
    depInfos.forEach((depInfo) => {
      const subtree = buildTreeRecurisve(
        depInfo.Name, new Set(), depsMap, virtualDepsMap, depNamesToSkip);
      if (subtree) {
        root.dependencies[subtree.name] = subtree;
      }
    });
  };

  // attach (as direct deps) pkgs not marked auto-installed:
  const manuallyInstalledDeps = depInfosList.filter((depInfo) => {
    return !depInfo.AutoInstalled;
  });
  attachDeps(manuallyInstalledDeps);

  // attach (as direct deps) pkgs marked as auto-insatalled,
  //  but not dependant upon:
  const notVisitedDeps = depInfosList.filter((depInfo) => {
    const depName = depInfo.Name;
    return !(depsMap[depName]._visited);
  });
  attachDeps(notVisitedDeps);

  // group all the "too frequest" deps under a meta package:
  if (tooFrequentDepNames.length > 0) {
    const tooFrequentDeps = tooFrequentDepNames.map((name) => {
      return depsMap[name];
    });

    const metaSubtree = {
      name: 'meta-common-packages',
      version: 'meta',
      dependencies: {},
    };

    tooFrequentDeps.forEach((depInfo) => {
      const pkg = {
        name: depFullName(depInfo),
        version: depInfo.Version,
      };
      metaSubtree.dependencies[pkg.name] = pkg;
    });

    root.dependencies[metaSubtree.name] = metaSubtree;
  }

  return root;
}

function buildTreeRecurisve(
  depName, ancestors, depsMap, virtualDepsMap, depNamesToSkip) {
  const depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return null;
  }

  // "realName" as the argument depName might be a virtual pkg
  const realName = depInfo.Name;
  const fullName = depFullName(depInfo);
  if (ancestors.has(fullName) || depNamesToSkip.has(realName)) {
    return null;
  }

  const tree: {
    name: string;
    version: string;
    dependencies?: any;
  } = {
    name: fullName,
    version: depInfo.Version,
  };

  if (depInfo._visited) {
    return tree;
  }
  depInfo._visited = true;

  const newAncestors = (new Set(ancestors)).add(fullName);

  const deps = depInfo.Deps || {};
  Object.keys(deps).forEach((name) => {
    const subTree = buildTreeRecurisve(
      name, newAncestors, depsMap, virtualDepsMap, depNamesToSkip);
    if (subTree) {
      if (!tree.dependencies) {
        tree.dependencies = {};
      }
      tree.dependencies[subTree.name] = subTree;
    }
  });

  return tree;
}

function countDepsRecursive(
  depName, ancestors, depsMap, virtualDepsMap, depCounts) {
  const depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return;
  }

  // "realName" as the argument depName might be a virtual pkg
  const realName = depInfo.Name;
  if (ancestors.has(realName)) {
    return;
  }

  depCounts[realName] = (depCounts[realName] || 0) + 1;

  const newAncestors = (new Set(ancestors)).add(realName);
  const deps = depInfo.Deps || {};
  Object.keys(deps).forEach((name) => {
    countDepsRecursive(
      name, newAncestors, depsMap, virtualDepsMap, depCounts);
  });
}

function depFullName(depInfo) {
  let fullName = depInfo.Name;
  if (depInfo.Source) {
    fullName = depInfo.Source + '/' + fullName;
  }
  return fullName;
}
