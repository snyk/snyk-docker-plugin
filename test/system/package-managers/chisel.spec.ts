import { scan } from "../../../lib/index";
import { execute } from "../../../lib/sub-process";

describe("chisel package manager tests", () => {
  afterAll(async () => {
    await execute("docker", [
      "image",
      "rm",
      "snykgoof/dockerhub-goof:ubuntu-chisel-24.04",
    ]).catch(() => {
      console.error(`tests teardown failed to remove docker image`);
    });
  });

  it("should correctly analyze a chiseled Ubuntu image", async () => {
    const image = "snykgoof/dockerhub-goof:ubuntu-chisel-24.04";
    const pluginResult = await scan({
      path: image,
      platform: "linux/amd64",
    });

    expect(pluginResult).toMatchSnapshot();
  });
});
