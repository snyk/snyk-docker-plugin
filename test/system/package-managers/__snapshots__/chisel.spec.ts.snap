// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chisel package manager tests should correctly analyze a chiseled Ubuntu image 1`] = `
Object {
  "scanResults": Array [
    Object {
      "facts": Array [
        Object {
          "data": Object {
            "graph": Object {
              "nodes": Array [
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "base-files@13ubuntu10.3",
                    },
                    Object {
                      "nodeId": "coreutils@9.4-3ubuntu6.1",
                    },
                    Object {
                      "nodeId": "dash@0.5.12-6ubuntu5",
                    },
                    Object {
                      "nodeId": "gcc-14-base@14.2.0-4ubuntu2~24.04",
                    },
                    Object {
                      "nodeId": "libacl1@2.3.2-1build1.1",
                    },
                    Object {
                      "nodeId": "libattr1@1:2.5.2-1build1.1",
                    },
                    Object {
                      "nodeId": "libaudit-common@1:3.1.2-2.1build1.1",
                    },
                    Object {
                      "nodeId": "libaudit1@1:3.1.2-2.1build1.1",
                    },
                    Object {
                      "nodeId": "libc6@2.39-0ubuntu8.6",
                    },
                    Object {
                      "nodeId": "libcap-ng0@0.8.4-2build2",
                    },
                    Object {
                      "nodeId": "libcap2@1:2.66-5ubuntu2.2",
                    },
                    Object {
                      "nodeId": "libcrypt1@1:4.4.36-4build1",
                    },
                    Object {
                      "nodeId": "libgcc-s1@14.2.0-4ubuntu2~24.04",
                    },
                    Object {
                      "nodeId": "libgmp10@2:6.3.0+dfsg-2ubuntu6.1",
                    },
                    Object {
                      "nodeId": "libpam0g@1.5.3-5ubuntu5.5",
                    },
                    Object {
                      "nodeId": "libpcre2-8-0@10.42-4ubuntu2.1",
                    },
                    Object {
                      "nodeId": "libselinux1@3.5-2ubuntu2.1",
                    },
                    Object {
                      "nodeId": "libssl3t64@3.0.13-0ubuntu3.6",
                    },
                    Object {
                      "nodeId": "tar@1.35+dfsg-3build1",
                    },
                  ],
                  "nodeId": "root-node",
                  "pkgId": "docker-image|snykgoof/dockerhub-goof@ubuntu-chisel-24.04",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "base-files@13ubuntu10.3",
                  "pkgId": "base-files@13ubuntu10.3",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "coreutils@9.4-3ubuntu6.1",
                  "pkgId": "coreutils@9.4-3ubuntu6.1",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "dash@0.5.12-6ubuntu5",
                  "pkgId": "dash@0.5.12-6ubuntu5",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "gcc-14-base@14.2.0-4ubuntu2~24.04",
                  "pkgId": "gcc-14-base@14.2.0-4ubuntu2~24.04",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libacl1@2.3.2-1build1.1",
                  "pkgId": "libacl1@2.3.2-1build1.1",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libattr1@1:2.5.2-1build1.1",
                  "pkgId": "libattr1@1:2.5.2-1build1.1",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libaudit-common@1:3.1.2-2.1build1.1",
                  "pkgId": "libaudit-common@1:3.1.2-2.1build1.1",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libaudit1@1:3.1.2-2.1build1.1",
                  "pkgId": "libaudit1@1:3.1.2-2.1build1.1",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libc6@2.39-0ubuntu8.6",
                  "pkgId": "libc6@2.39-0ubuntu8.6",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libcap-ng0@0.8.4-2build2",
                  "pkgId": "libcap-ng0@0.8.4-2build2",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libcap2@1:2.66-5ubuntu2.2",
                  "pkgId": "libcap2@1:2.66-5ubuntu2.2",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libcrypt1@1:4.4.36-4build1",
                  "pkgId": "libcrypt1@1:4.4.36-4build1",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libgcc-s1@14.2.0-4ubuntu2~24.04",
                  "pkgId": "libgcc-s1@14.2.0-4ubuntu2~24.04",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libgmp10@2:6.3.0+dfsg-2ubuntu6.1",
                  "pkgId": "libgmp10@2:6.3.0+dfsg-2ubuntu6.1",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libpam0g@1.5.3-5ubuntu5.5",
                  "pkgId": "libpam0g@1.5.3-5ubuntu5.5",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libpcre2-8-0@10.42-4ubuntu2.1",
                  "pkgId": "libpcre2-8-0@10.42-4ubuntu2.1",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libselinux1@3.5-2ubuntu2.1",
                  "pkgId": "libselinux1@3.5-2ubuntu2.1",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libssl3t64@3.0.13-0ubuntu3.6",
                  "pkgId": "libssl3t64@3.0.13-0ubuntu3.6",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "tar@1.35+dfsg-3build1",
                  "pkgId": "tar@1.35+dfsg-3build1",
                },
              ],
              "rootNodeId": "root-node",
            },
            "pkgManager": Object {
              "name": "deb",
              "repositories": Array [
                Object {
                  "alias": "ubuntu:24.04",
                },
              ],
            },
            "pkgs": Array [
              Object {
                "id": "docker-image|snykgoof/dockerhub-goof@ubuntu-chisel-24.04",
                "info": Object {
                  "name": "docker-image|snykgoof/dockerhub-goof",
                  "version": "ubuntu-chisel-24.04",
                },
              },
              Object {
                "id": "base-files@13ubuntu10.3",
                "info": Object {
                  "name": "base-files",
                  "version": "13ubuntu10.3",
                },
              },
              Object {
                "id": "coreutils@9.4-3ubuntu6.1",
                "info": Object {
                  "name": "coreutils",
                  "version": "9.4-3ubuntu6.1",
                },
              },
              Object {
                "id": "dash@0.5.12-6ubuntu5",
                "info": Object {
                  "name": "dash",
                  "version": "0.5.12-6ubuntu5",
                },
              },
              Object {
                "id": "gcc-14-base@14.2.0-4ubuntu2~24.04",
                "info": Object {
                  "name": "gcc-14-base",
                  "version": "14.2.0-4ubuntu2~24.04",
                },
              },
              Object {
                "id": "libacl1@2.3.2-1build1.1",
                "info": Object {
                  "name": "libacl1",
                  "version": "2.3.2-1build1.1",
                },
              },
              Object {
                "id": "libattr1@1:2.5.2-1build1.1",
                "info": Object {
                  "name": "libattr1",
                  "version": "1:2.5.2-1build1.1",
                },
              },
              Object {
                "id": "libaudit-common@1:3.1.2-2.1build1.1",
                "info": Object {
                  "name": "libaudit-common",
                  "version": "1:3.1.2-2.1build1.1",
                },
              },
              Object {
                "id": "libaudit1@1:3.1.2-2.1build1.1",
                "info": Object {
                  "name": "libaudit1",
                  "version": "1:3.1.2-2.1build1.1",
                },
              },
              Object {
                "id": "libc6@2.39-0ubuntu8.6",
                "info": Object {
                  "name": "libc6",
                  "version": "2.39-0ubuntu8.6",
                },
              },
              Object {
                "id": "libcap-ng0@0.8.4-2build2",
                "info": Object {
                  "name": "libcap-ng0",
                  "version": "0.8.4-2build2",
                },
              },
              Object {
                "id": "libcap2@1:2.66-5ubuntu2.2",
                "info": Object {
                  "name": "libcap2",
                  "version": "1:2.66-5ubuntu2.2",
                },
              },
              Object {
                "id": "libcrypt1@1:4.4.36-4build1",
                "info": Object {
                  "name": "libcrypt1",
                  "version": "1:4.4.36-4build1",
                },
              },
              Object {
                "id": "libgcc-s1@14.2.0-4ubuntu2~24.04",
                "info": Object {
                  "name": "libgcc-s1",
                  "version": "14.2.0-4ubuntu2~24.04",
                },
              },
              Object {
                "id": "libgmp10@2:6.3.0+dfsg-2ubuntu6.1",
                "info": Object {
                  "name": "libgmp10",
                  "version": "2:6.3.0+dfsg-2ubuntu6.1",
                },
              },
              Object {
                "id": "libpam0g@1.5.3-5ubuntu5.5",
                "info": Object {
                  "name": "libpam0g",
                  "version": "1.5.3-5ubuntu5.5",
                },
              },
              Object {
                "id": "libpcre2-8-0@10.42-4ubuntu2.1",
                "info": Object {
                  "name": "libpcre2-8-0",
                  "version": "10.42-4ubuntu2.1",
                },
              },
              Object {
                "id": "libselinux1@3.5-2ubuntu2.1",
                "info": Object {
                  "name": "libselinux1",
                  "version": "3.5-2ubuntu2.1",
                },
              },
              Object {
                "id": "libssl3t64@3.0.13-0ubuntu3.6",
                "info": Object {
                  "name": "libssl3t64",
                  "version": "3.0.13-0ubuntu3.6",
                },
              },
              Object {
                "id": "tar@1.35+dfsg-3build1",
                "info": Object {
                  "name": "tar",
                  "version": "1.35+dfsg-3build1",
                },
              },
            ],
            "schemaVersion": "1.3.0",
          },
          "type": "depGraph",
        },
        Object {
          "data": "sha256:5262b11cd2c465ec76e4f8dc00522e7065c428786e7ee0f2f91ef59eed343dc0",
          "type": "imageId",
        },
        Object {
          "data": Array [
            "sha256:214740def7356fff18e642236c22109b81bf104c163f3c055b974e37682f67c1",
          ],
          "type": "imageLayers",
        },
        Object {
          "data": "2025-10-22T17:28:05.980889584Z",
          "type": "imageCreationTime",
        },
        Object {
          "data": Array [
            "sha256:90024d52c80c098e7ad0c4dfd4d2d7c3b9709d5bd2efe023b2bde587fa3ce139",
          ],
          "type": "rootFs",
        },
        Object {
          "data": "Ubuntu 24.04.3 LTS",
          "type": "imageOsReleasePrettyName",
        },
        Object {
          "data": Object {
            "names": Array [
              "snykgoof/dockerhub-goof:ubuntu-chisel-24.04",
               "snykgoof/dockerhub-goof@sha256:6c33709164fe0c9bc20e3d13ba3c1f2fcd6fc7fcf6357b84fbc1c50c483a1178",
               "snykgoof/dockerhub-goof@sha256:2f7f8c4bb123b101e4356a8955cf294c0a7354249e5d8cc7b8a4e7bb02271cbc",
            ],
          },
          "type": "imageNames",
        },
      ],
      "identity": Object {
        "args": Object {
          "platform": "linux/amd64",
        },
        "type": "deb",
      },
      "target": Object {
        "image": "docker-image|snykgoof/dockerhub-goof",
      },
    },
  ],
}
`;
